### PR TITLE
Switch to a non noop logger with trace ids

### LINKED ISSUE #118: Server errors don't actually log
Looking at https://github.com/grpc-ecosystem/go-grpc-middleware/blob/master/logging/logrus/ctxlogrus/context.go#L36 it looks like you need to call ctxlogrus.ToContext(context) first, before you can call extract.

I can't find ToContext being called. This makes https://github.com/freifunkMUC/wg-access-server/blob/master/internal/services/middleware.go#L22 a null logger.

### PR DESCRIPTION
**Pull request by halkeye**
Fixes #118

### CODE REVIEW

File: internal/services/middleware.go [hunk 1]
```diff
@@ -19,7 +20,13 @@ func RecoveryMiddleware(next http.Handler) http.Handler {
 	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
 		defer func() {
 			if err := recover(); err != nil {
-				ctxlogrus.Extract(r.Context()).
+				logger := ctxlogrus.Extract(r.Context())
+
+				if logger.Logger.Out == ioutil.Discard {
```
Comment (halkeye) [inline internal/services/middleware.go:25]: not sold on this check, but atm that's the only way to detect the null logger. 

https://github.com/grpc-ecosystem/go-grpc-middleware/issues/467

Is there a way to tell if its an API hit or not?

**Comment by halkeye** [comment]
```
ERRO[0087]middleware.go:31 runtime error: invalid memory address or nil pointer dereference  stack="goroutine 68 [running]:\nruntime/debug.Stack()\n\t/usr/local/go/src/runtime/debug/stack.go:24 +0x65\ngithub.com/freifunkMUC/wg-access-server/internal/services.RecoveryMiddleware.func1.1()\n\t/home/halkeye/go/src/github.com/freifunkMUC/wg-access-server/internal/services/middleware.go:30 +0x107\npanic({0xca1da0, 0x13ce6a0})\n\t/usr/local/go/src/runtime/panic.go:1038 +0x215\ngithub.com/freifunkMUC/wg-access-server/cmd/serve.(*servecmd).Run.func5(0xea81e8)\n\t/home/halkeye/go/src/github.com/freifunkMUC/wg-access-server/cmd/serve/main.go:340 +0x18\ngithub.com/freifunkMUC/wg-access-server/pkg/authnz.(*AuthMiddleware).Middleware.func1({0xea87e8, 0xc0005a2d20}, 0xc000324300)\n\t/home/halkeye/go/src/github.com/freifunkMUC/wg-access-server/pkg/authnz/router.go:97 +0xd9\nnet/http.HandlerFunc.ServeHTTP(0x100000000c63aa0, {0xea87e8, 0xc0005a2d20}, 0x90)\n\t/usr/local/go/src/net/http/server.go:2047 +0x2f\ngithub.com/freifunkMUC/wg-access-server/internal/services.RecoveryMiddleware.func1({0xea87e8, 0xc0005a2d20}, 0xc000175901)\n\t/home/halkeye/go/src/github.com/freifunkMUC/wg-access-server/internal/services/middleware.go:36 +0x83\nnet/http.HandlerFunc.ServeHTTP(0xead300, {0xea87e8, 0xc0005a2d20}, 0xc000175988)\n\t/usr/local/go/src/net/http/server.go:2047 +0x2f\ngithub.com/freifunkMUC/wg-access-server/internal/services.TracesMiddleware.func1({0xea87e8, 0xc0005a2d20}, 0xc000324200)\n\t/home/halkeye/go/src/github.com/freifunkMUC/wg-access-server/internal/services/middleware.go:15 +0x230\nnet/http.HandlerFunc.ServeHTTP(0xc000324100, {0xea87e8, 0xc0005a2d20}, 0xc0001759f8)\n\t/usr/local/go/src/net/http/server.go:2047 +0x2f\ngithub.com/gorilla/mux.(*Router).ServeHTTP(0xc0000d2d80, {0xea87e8, 0xc0005a2d20}, 0xc000324000)\n\t/home/halkeye/go/pkg/mod/github.com/gorilla/mux@v1.8.0/mux.go:210 +0x1cf\nnet/http.serverHandler.ServeHTTP({0xc0003a43c0}, {0xea87e8, 0xc0005a2d20}, 0xc000324000)\n\t/usr/local/go/src/net/http/server.go:2879 +0x43b\nnet/http.(*conn).serve(0xc0000e40a0, {0xead300, 0xc0003a42a0})\n\t/usr/local/go/src/net/http/server.go:1930 +0xb08\ncreated by net/http.(*Server).Serve\n\t/usr/local/go/src/net/http/server.go:3034 +0x4e8\n" trace.id=6db428cd-4fe0-4023-ad74-336bcca427a3```

Might be nice later to include the route but at least now its not silent.

File: internal/services/middleware.go (hunk 1 above)
Comment (DasSkelett) [reply internal/services/middleware.go:25]: Maybe this?
```go
isGrpc := (&grpcweb.WrappedGrpcServer{}).IsGrpcWebRequest(r)
if isGrpc {
	logger = ctxlogrus.Extract(r.Context())
} else {
	logger = traces.Logger(r.Context())
}
```
https://pkg.go.dev/github.com/improbable-eng/grpc-web/go/grpcweb@v0.15.0#WrappedGrpcServer.IsGrpcWebRequest
Still ugly, but maybe a bit better.

**Comment by DasSkelett** [comment]
While playing around I found out that this doesn't actually catch panics in grpc code anyway, because it handles the requests in separate coroutines;
https://github.com/grpc/grpc-go/blob/431ea809a7676e1da8d09c33ae0d31fcba85f1ff/server.go#L920-L923
(Found the code through the stacktrace of the unrecovered panic)

Instead, we should probably use the [`grpc_recovery` interceptor](https://pkg.go.dev/github.com/grpc-ecosystem/go-grpc-middleware@v1.3.0/recovery) for those:

```go
func ApiRouter(deps *ApiServices) http.Handler {
	// Native GRPC server
	server := grpc.NewServer([]grpc.ServerOption{
		grpc.MaxRecvMsgSize(int(1 * math.Pow(2, 20))), // 1MB
		grpc.UnaryInterceptor(grpc_middleware.ChainUnaryServer(
			func(ctx context.Context, req interface{}, info *grpc.UnaryServerInfo, handler grpc.UnaryHandler) (resp interface{}, err error) {
				return grpc_logrus.UnaryServerInterceptor(traces.Logger(ctx))(ctx, req, info, handler)
			},
			grpc_recovery.UnaryServerInterceptor(),
		)),
	}...)
```
(From https://github.com/freifunkMUC/wg-access-server/blob/e7c215165c15b76de5bc75ce07d39c01be036b54/internal/services/api_router.go#L26-L33)

With that addition it does indeed recover and log the panic correctly:
```
ERRO[0005]options.go:224 finished unary call with code Internal        error="rpc error: code = Internal desc = test error" grpc.code=Internal grpc.method=ListDevices grpc.service=proto.Devices grpc.start_time="2022-03-05T20:17:48+01:00" grpc.time_ms=0.041 span.kind=server system=grpc trace.id=cff269dc-c8a1-4e67-9095-e38362f93eca
```

And we could get rid of the ctxlogrus special handling in the `RecoveryMiddleware` and only use the `traces.Logger`.

**Comment by halkeye** [comment]
the lack of tests is really frustrating. I've not done any low level grpc stuff yet (we have a framework at work). How hard would it be to setup a grpc test? I can write them, but need help with the scaffolding

**Comment by DasSkelett** [comment]
Yup, this project really could use more tests.

> How hard would it be to setup a grpc test? I can write them, but need help with the scaffolding

Honestly no idea. I haven't worked with grpc at all before. I only know the basics of protobuf, and a handful of things I've seen in this repository.

**Comment by halkeye** [comment]
I won't be able to look at this for a couple hours, but I don't want to lose https://stackoverflow.com/questions/42102496/testing-a-grpc-service link

**Comment by elcomtik** [comment]
Is this ready to merge?

**Comment by DasSkelett** [comment]
> Is this ready to merge?

No @elcomtik, if it were ready to be merged it would already be merged. I can understand your desire to make use of your own contribution, but impatiently spamming other PRs is rather impolite and doesn't help at all to speed things up.

**Comment by elcomtik** [comment]
I'm sorry, will not happen again.

**Comment by DasSkelett** [comment]
I've split the logging and recovery fixes to #155 now

**Comment by GoliathLabs** [comment]
@halkeye there are some merge conflicts that need to be resolved. In the meantime, what do you think of the approach proposed by @DasSkelett in https://github.com/freifunkMUC/wg-access-server/pull/119#discussion_r820136153?

**Comment by halkeye** [comment]
sounds reasonable. I'll admit I got sidetracked and haven't looked into this in a while.

**Comment by GoliathLabs** [comment]
@halkeye can you resolve the merge conflict?

**Comment by halkeye** [comment]
probably that? its been two months since i poked at this sorry.

**Comment by halkeye** [comment]
nope, i'll have to pull it down locally to fix it :(
I'll see if i remember to do it tomorrow

**Comment by halkeye** [comment]
Honestly now that I have the system working I havnt really spent time thinking about it. I can try looking at it tonight after work, but also fine closing it

**Comment by DasSkelett** [comment]
Closing now, feel free to reopen this or a new one when you come back to the grpc tests.

**Pull request closed without merging**